### PR TITLE
feat(android, sdks)!: update to the latest gms:play-services-ads:20.5.0

### PR DESCRIPTION
**Pull request by Mikenso**
### Description

Android Update Banner, Interstitial, Rewarded add to be compatible with com.google.android.gms:play-services-ads:20.5.0

BREAKING CHANGES:
Please refer to upstream guides for suggestions on new usage. https://developers.google.com/admob/ios/migration and https://developers.google.com/admob/android/migration
- compileSdkVersion now 31, change your app android build.gradle to 31 if you have not already. Note that JDK11 is required for stable compilation on compileSdkVersion 31, JDK8 has internal compiler errors with SDK31
- onAdLeftApplication removed from the underlying SDK, use react-native built in AppState to determine app went to background
- Smart banner ads removed; use adaptive banner ads. Set height/width explicitly taking into account device size

### CODE REVIEW

**Comment by mikehardy** [comment]
Oh fantastic! That was next up on the todo list and it appears you have nailed it! Is this a breaking change? I did not see any changes to the javsascript layer so this appears to be all on the native side. Is that a correct guess? I have not reviewed though - just skimmed really quickly before approving the CI run

Can you sign the CLA agreement as well? Excited to move quickly on this one, and I really appreciate that you posted this

**Comment by RodolfoGS** [comment]
`onAdLeftApplication` will be removed? So, we'll not be able to know when a user left the app after tapped on an ad?

**Comment by mikehardy** [comment]
@RodolfoGS check the underlying SDK docs for reference https://developers.google.com/admob/android/migration
what do they say on the matter?

**Comment by RodolfoGS** [comment]
Mmm yes, I see, _Leave application callback removal_. Ok, thanks for the clarification!

**Comment by mikehardy** [comment]
no problem - I have not read it much myself so I was not actually sure, but I had the link handy ;-). Cheers

**Comment by mikehardy** [comment]
I marked this as a breaking change (with a title update on the PR including `!`) and that means we need to generate an excellent note for library consumers indicating what the breaking change(s) is/are, and how they need to react.

It seems like this one might simply be:

BREAKING CHANGE: onAdLeftApplication is removed from the underlying SDK, there is no replacement

**Comment by mikehardy** [comment]
@Mikenso also needs a bump to compileSdkVersion to 31:

```
Execution failed for task ':app:checkDebugAarMetadata'.
> A failure occurred while executing com.android.build.gradle.internal.tasks.CheckAarMetadataWorkAction
   > The minCompileSdk (31) specified in a
     dependency's AAR metadata (META-INF/com/android/build/gradle/aar-metadata.properties)
     is greater than this module's compileSdkVersion (android-30).
     Dependency: androidx.work:work-runtime:2.7.0.
     AAR metadata file: /Users/runner/.gradle/caches/transforms-3/4739effbb596877ca3318d6c0a41cf71/transformed/work-runtime-2.7.0/META-INF/com/android/build/gradle/aar-metadata.properties.
```

BREAKING CHANGE: compileSdkVersion now 31, change your app android build.gradle to 31 if you have not already

**Comment by mikehardy** [comment]
@Mikenso run google java formatter locally and commit the result, it will fix the lint errors:

```

**Review by mikehardy — CHANGES_REQUESTED** [review]
- lint errors
- example app update to compileSdkVersion 31
- needs a crystal clear description of all breaking changes, to help app developers deal with the changes

**Comment by dylancom** [comment]
I think that also here setting testDevices has to move to the overall configuration instead of per ad. And so it has to be set in: https://github.com/invertase/react-native-google-ads/blob/9569b98768cacd3f0d3e95b7b44d15d20e59681a/android/src/main/java/io/invertase/googleads/ReactNativeGoogleAdsModule.java

Besides that the key "testDevices" became: "testDeviceIdentifiers".

So this PR should match with setting test devices as follow:
`admob().setRequestConfiguration({testDeviceIdentifiers: ["EMULATOR"]});`

**Comment by mikehardy** [comment]
Agreed with Dylan's comment on making sure the test device stuff is in harmony with iOS changes as mentioned, thanks for the review Dylan

**Comment by Mikenso** [comment]
Hi @mikehardy, @dylancom, thanks for review

Where do I need to specify breaking changes?
It would be:

### BREAKING CHANGES:
- compileSdkVersion now 31, change your app android build.gradle to 31 if you have not already. Note that JDK11 is required for stable compilation on compileSdkVersion 31, JDK8 has internal compiler errors with SDK31
- onAdLeftApplication is removed from the underlying SDK, there is no replacement
- Smart banner ads are deprecated in favor of adaptive banner ads.

File: android/src/main/java/io/invertase/googleads/ReactNativeGoogleAdsBannerAdViewManager.java [hunk 1]
```diff
@@ -31,16 +32,17 @@
 import com.google.android.gms.ads.AdRequest;
 import com.google.android.gms.ads.AdSize;
 import com.google.android.gms.ads.AdView;
+import com.google.android.gms.ads.LoadAdError;
 import java.util.Map;
 import javax.annotation.Nonnull;
+import org.jetbrains.annotations.NotNull;
```
Comment (mikehardy) [inline android/src/main/java/io/invertase/googleads/ReactNativeGoogleAdsBannerAdViewManager.java:38]: Please use the androidx annotations, you already have `androidx-annotation.NonNull` above, no need for jetbrains NotNull as well 
```suggestion
```

File: android/src/main/java/io/invertase/googleads/ReactNativeGoogleAdsBannerAdViewManager.java [hunk 2]
```diff
@@ -160,7 +155,8 @@ public void onAdLoaded() {
           }
 
           @Override
-          public void onAdFailedToLoad(int errorCode) {
+          public void onAdFailedToLoad(@NonNull @NotNull LoadAdError loadAdError) {
```
Comment (mikehardy) [inline android/src/main/java/io/invertase/googleads/ReactNativeGoogleAdsBannerAdViewManager.java:158]:
```suggestion
          public void onAdFailedToLoad(@NonNull LoadAdError loadAdError) {
```

File: android/src/main/java/io/invertase/googleads/ReactNativeGoogleAdsInterstitialModule.java [hunk 3]
```diff
@@ -17,28 +17,32 @@
...
-import com.google.android.gms.ads.AdListener;
-import com.google.android.gms.ads.InterstitialAd;
+import com.google.android.gms.ads.AdRequest;
+import com.google.android.gms.ads.FullScreenContentCallback;
+import com.google.android.gms.ads.LoadAdError;
+import com.google.android.gms.ads.interstitial.InterstitialAd;
+import com.google.android.gms.ads.interstitial.InterstitialAdLoadCallback;
 import io.invertase.googleads.common.ReactNativeModule;
 import javax.annotation.Nullable;
+import org.jetbrains.annotations.NotNull;
```
Comment (mikehardy) [inline android/src/main/java/io/invertase/googleads/ReactNativeGoogleAdsInterstitialModule.java:45]: same - stick with androidx annotations
```suggestion
```

File: android/src/main/java/io/invertase/googleads/ReactNativeGoogleAdsInterstitialModule.java [hunk 4]
```diff
@@ -67,51 +70,51 @@ public void interstitialLoad(int requestId, String adUnitId, ReadableMap adReque
...
-          // Apply AdRequest builder
-          interstitialAd.loadAd(buildAdRequest(adRequestOptions));
+          InterstitialAdLoadCallback interstitialAdLoadCallback =
+              new InterstitialAdLoadCallback() {
 
-          interstitialAd.setAdListener(
-              new AdListener() {
                 @Override
-                public void onAdLoaded() {
+                public void onAdLoaded(@NonNull @NotNull InterstitialAd interstitialAd) {
```
Comment (mikehardy) [inline android/src/main/java/io/invertase/googleads/ReactNativeGoogleAdsInterstitialModule.java:80]:
```suggestion
                public void onAdLoaded(@NonNull InterstitialAd interstitialAd) {
```

File: android/src/main/java/io/invertase/googleads/ReactNativeGoogleAdsInterstitialModule.java [hunk 5]
```diff
@@ -67,51 +70,51 @@ public void interstitialLoad(int requestId, String adUnitId, ReadableMap adReque
...
+                        }
+                      });
+
+                  interstitialAdArray.put(requestId, interstitialAd);
                   sendInterstitialEvent(GOOGLE_ADS_EVENT_LOADED, requestId, adUnitId, null);
                 }
 
                 @Override
-                public void onAdFailedToLoad(int errorCode) {
+                public void onAdFailedToLoad(@NonNull @NotNull LoadAdError loadAdError) {
```
Comment (mikehardy) [inline android/src/main/java/io/invertase/googleads/ReactNativeGoogleAdsInterstitialModule.java:107]:
```suggestion
                public void onAdFailedToLoad(@NonNull LoadAdError loadAdError) {
```

**Review by mikehardy — CHANGES_REQUESTED** [review]
Starting to look good - I still need to do a deeper look (in progress) but here's the results of a surface level scan - simple item of annotation style

**Comment by mikehardy** [comment]
Looks like it needs either `yarn lint:prettier:fix` or `yarn:eslint:fix` as well. The lint tasks need to pass

```

$ yarn lint:js && yarn lint:android && yarn lint:ios:check
$ eslint lib/ --ext .js,.jsx,.ts,.tsx --max-warnings=0

/home/runner/work/react-native-google-ads/react-native-google-ads/lib/index.d.ts
Error:   123:24  error  Delete `····`  prettier/prettier

✖ 1 problem (1 error, 0 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.
```

**Review by mikehardy — COMMENTED** [review]
looked good - I confirmed smart banners were gone on ios side as well, and left-app event was also gone on ios, so API is now consistent across platforms with both up to date on SDKs
This looks like a solid forward port, thank you

**Pull request merged**